### PR TITLE
Fix segfault in groupConcat

### LINKED ISSUE #65426: Segfault in `AggregateFunctionGroupConcat.cpp:193`
https://s3.amazonaws.com/clickhouse-test-reports/65239/ece3efe09cc584f3ac0320e83e3829684153c1d9/stress_test__debug_.html

query: `SELECT groupConcat(p_int) FROM test_groupConcat`
```
2024.06.18 21:15:50.140805 [ 15147 ] {} <Fatal> BaseDaemon: ########## Short fault info ############
2024.06.18 21:15:50.145900 [ 15147 ] {} <Fatal> BaseDaemon: (version 24.6.1.4137, build id: D160E7D4D9D03A27F7E62BAE039D3F3ADC9E0467, git hash: 3d440d5d6b60fdd47b31bad277c9aeb8d25e2ff9) (from thread 13005) Received signal 11
2024.06.18 21:15:50.147155 [ 15147 ] {} <Fatal> BaseDaemon: Signal description: Segmentation fault
2024.06.18 21:15:50.148160 [ 15147 ] {} <Fatal> BaseDaemon: Address: 0x1. Access: write. Address not mapped to object.
2024.06.18 21:15:50.149914 [ 15147 ] {} <Fatal> BaseDaemon: Stack trace: 0x00000000130e374a 0x00000000130e3640 0x000000001350a831 0x00007fd64afc0520 0x00000000268ed634 0x00000000268ed545 0x000000000b14f9ac 0x0000000013159c68 0x0000000013cfdbe5 0x0000000015e08ae1 0x000000001a3f9e9a 0x000000001a3ea8ba 0x000000001d912296 0x000000001d9131a0 0x000000001d711efe 0x000000001d711a28 0x000000001d711371 0x000000001d779958 0x000000001a2d2a91 0x000000001a2d15c9 0x000000001a2d14d8 0x000000001a2d1468 0x000000001a2d1428 0x000000001a2d135e 0x000000001a2d107d
2024.06.18 21:15:50.152028 [ 15147 ] {} <Fatal> BaseDaemon: ########################################
2024.06.18 21:15:50.161079 [ 15147 ] {} <Fatal> BaseDaemon: (version 24.6.1.4137, build id: D160E7D4D9D03A27F7E62BAE039D3F3ADC9E0467, git hash: 3d440d5d6b60fdd47b31bad277c9aeb8d25e2ff9) (from thread 13005) (query_id: 708401fe-5db0-43e9-9801-862371b114b2) (query: SELECT groupConcat(p_int) FROM test_groupConcat;) Received signal Segmentation fault (11)
2024.06.18 21:15:50.175099 [ 15147 ] {} <Fatal> BaseDaemon: Address: 0x1. Access: write. Address not mapped to object.
2024.06.18 21:15:50.196934 [ 15147 ] {} <Fatal> BaseDaemon: Stack trace: 0x00000000130e374a 0x00000000130e3640 0x000000001350a831 0x00007fd64afc0520 0x00000000268ed634 0x00000000268ed545 0x000000000b14f9ac 0x0000000013159c68 0x0000000013cfdbe5 0x0000000015e08ae1 0x000000001a3f9e9a 0x000000001a3ea8ba 0x000000001d912296 0x000000001d9131a0 0x000000001d711efe 0x000000001d711a28 0x000000001d711371 0x000000001d779958 0x000000001a2d2a91 0x000000001a2d15c9 0x000000001a2d14d8 0x000000001a2d1468 0x000000001a2d1428 0x000000001a2d135e 0x000000001a2d107d
2024.06.18 21:15:50.374388 [ 15147 ] {} <Fatal> BaseDaemon: 0. /build/src/Common/StackTrace.cpp:349: StackTrace::tryCapture() @ 0x00000000130e374a
2024.06.18 21:15:50.680279 [ 15147 ] {} <Fatal> BaseDaemon: 1. /build/src/Common/StackTrace.cpp:323: StackTrace::StackTrace(ucontext_t const&) @ 0x00000000130e3640
2024.06.18 21:15:51.137841 [ 15147 ] {} <Fatal> BaseDaemon: 2. /build/src/Daemon/BaseDaemon.cpp:158: signalHandler(int, siginfo_t*, void*) @ 0x000000001350a831
2024.06.18 21:15:51.145260 [ 15147 ] {} <Fatal> BaseDaemon: 3. ? @ 0x00007fd64afc0520
2024.06.18 21:15:51.233016 [ 15147 ] {} <Fatal> BaseDaemon: 4. /build/base/glibc-compatibility/memcpy/memcpy.h:127: inline_memcpy(void*, void const*, unsigned long) @ 0x00000000268ed634
2024.06.18 21:15:51.259755 [ 15147 ] {} <Fatal> BaseDaemon: 5. /build/base/glibc-compatibility/memcpy/memcpy.cpp:6: memcpy @ 0x00000000268ed545
2024.06.18 21:15:52.570077 [ 15147 ] {} <Fatal> BaseDaemon: 6. /build/src/IO/ReadBuffer.h:186: DB::ReadBuffer::read(char*, unsigned long) @ 0x000000000b14f9ac
2024.06.18 21:15:52.680891 [ 15147 ] {} <Fatal> BaseDaemon: 7. /build/src/IO/ReadBuffer.h:197: DB::ReadBuffer::readStrict(char*, unsigned long) @ 0x0000000013159c68
2024.06.18 21:15:53.101062 [ 15147 ] {} <Fatal> BaseDaemon: 8. /build/src/AggregateFunctions/AggregateFunctionGroupConcat.cpp:193: DB::(anonymous namespace)::GroupConcatImpl<false>::deserialize(char*, DB::ReadBuffer&, std::optional<unsigned long>, DB::Arena*) const @ 0x0000000013cfdbe5
2024.06.18 21:15:53.887451 [ 15147 ] {} <Fatal> BaseDaemon: 9. /build/src/AggregateFunctions/Combinators/AggregateFunctionNull.h:190: DB::AggregateFunctionNullBase<true, true, DB::AggregateFunctionNullUnary<true, true>>::deserialize(char*, DB::ReadBuffer&, std::optional<unsigned long>, DB::Arena*) const @ 0x0000000015e08ae1
2024.06.18 21:15:53.976340 [ 15147 ] {} <Fatal> BaseDaemon: 10. /build/src/DataTypes/Serializations/SerializationAggregateFunction.cpp:94: DB::SerializationAggregateFunction::deserializeBinaryBulk(DB::IColumn&, DB::ReadBuffer&, unsigned long, double) const @ 0x000000001a3f9e9a
2024.06.18 21:15:54.088859 [ 15147 ] {} <Fatal> BaseDaemon: 11. /build/src/DataTypes/Serializations/ISerialization.cpp:145: DB::ISerialization::deserializeBinaryBulkWithMultipleStreams(COW<DB::IColumn>::immutable_ptr<DB::IColumn>&, unsigned long, DB::ISerialization::DeserializeBinaryBulkSettings&, std::shared_ptr<DB::ISerialization::DeserializeBinaryBulkState>&, std::unordered_map<String, COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::hash<String>, std::equal_to<String>, std::allocator<std::pair<String const, COW<DB::IColumn>::immutable_ptr<DB::IColumn>>>>*) const @ 0x000000001a3ea8ba
2024.06.18 21:15:54.248546 [ 15147 ] {} <Fatal> BaseDaemon: 12. /build/src/Formats/NativeReader.cpp:97: DB::NativeReader::readData(DB::ISerialization const&, COW<DB::IColumn>::immutable_ptr<DB::IColumn>&, DB::ReadBuffer&, unsigned long, double) @ 0x000000001d912296
2024.06.18 21:15:54.409934 [ 15147 ] {} <Fatal> BaseDaemon: 13. /build/src/Formats/NativeReader.cpp:206: DB::NativeReader::read() @ 0x000000001d9131a0
2024.06.18 21:15:55.325776 [ 15147 ] {} <Fatal> BaseDaemon: 14. /build/src/Client/Connection.cpp:1175: DB::Connection::receiveDataImpl(DB::NativeReader&) @ 0x000000001d711efe
2024.06.18 21:15:56.370338 [ 15147 ] {} <Fatal> BaseDaemon: 15. /build/src/Client/Connection.cpp:1156: DB::Connection::receiveData() @ 0x000000001d711a28
2024.06.18 21:15:57.249517 [ 15147 ] {} <Fatal> BaseDaemon: 16. /build/src/Client/Connection.cpp:1082: DB::Connection::receivePacket() @ 0x000000001d711371
2024.06.18 21:15:58.164920 [ 15147 ] {} <Fatal> BaseDaemon: 17. /build/src/Client/MultiplexedConnections.cpp:333: DB::MultiplexedConnections::receivePacketUnlocked(std::function<void (int, Poco::Timespan, DB::AsyncEventTimeoutType, String const&, unsigned int)>) @ 0x000000001d779958
2024.06.18 21:15:58.451946 [ 15147 ] {} <Fatal> BaseDaemon: 18. /build/src/QueryPipeline/RemoteQueryExecutorReadContext.cpp:50: DB::RemoteQueryExecutorReadContext::Task::run(std::function<void (int, Poco::Timespan, DB::AsyncEventTimeoutType, String const&, unsigned int)>, std::function<void ()>) @ 0x000000001a2d2a91
2024.06.18 21:15:58.604807 [ 15147 ] {} <Fatal> BaseDaemon: 19. /build/src/Common/AsyncTaskExecutor.cpp:88: DB::AsyncTaskExecutor::Routine::operator()(std::function<void ()>) @ 0x000000001a2d15c9
2024.06.18 21:15:58.993834 [ 15147 ] {} <Fatal> BaseDaemon: 20. /build/src/Common/Fiber.h:76: Fiber::RoutineImpl<DB::AsyncTaskExecutor::Routine>::operator()(boost::context::fiber&&) @ 0x000000001a2d14d8
2024.06.18 21:15:59.284764 [ 15147 ] {} <Fatal> BaseDaemon: 21. /build/contrib/llvm-project/libcxx/include/__functional/invoke.h:394: ? @ 0x000000001a2d1468
2024.06.18 21:15:59.499838 [ 15147 ] {} <Fatal> BaseDaemon: 22. /build/contrib/llvm-project/libcxx/include/__functional/invoke.h:531: ? @ 0x000000001a2d1428
2024.06.18 21:15:59.678279 [ 15147 ] {} <Fatal> BaseDaemon: 23. /build/contrib/boost/boost/context/fiber_fcontext.hpp:157: boost::context::detail::fiber_record<boost::context::fiber, FiberStack&, Fiber::RoutineImpl<DB::AsyncTaskExecutor::Routine>>::run(void*) @ 0x000000001a2d135e
2024.06.18 21:15:59.972923 [ 15147 ] {} <Fatal> BaseDaemon: 24. /build/contrib/boost/boost/context/fiber_fcontext.hpp:97: void boost::context::detail::fiber_entry<boost::context::detail::fiber_record<boost::context::fiber, FiberStack&, Fiber::RoutineImpl<DB::AsyncTaskExecutor::Routine>>>(boost::context::detail::transfer_t) @ 0x000000001a2d107d
2024.06.18 21:15:59.997560 [ 15147 ] {} <Fatal> BaseDaemon: Integrity check of the executable skipped because the reference checksum could not be read.
2024.06.18 21:16:09.233320 [ 15147 ] {} <Fatal> BaseDaemon: This ClickHouse version is not official and should be upgraded to the official build.
2024.06.18 21:16:09.257731 [ 15147 ] {} <Fatal> BaseDaemon: Changed settings: min_compress_block_size = 2477463, max_compress_block_size = 1387141, max_block_size = 30829, min_insert_block_size_rows = 0, min_insert_block_size_bytes = 0, min_external_table_block_size_bytes = 100000000, max_joined_block_size_rows = 14304, max_insert_threads = 1, max_threads = 1, max_parsing_threads = 10, max_read_buffer_size = 605364, connect_timeout_with_failover_ms = 2000, connect_timeout_with_failover_secure_ms = 3000, idle_connection_timeout = 36000, s3_check_objects_after_upload = true, stream_like_engine_allow_direct_select = true, replication_wait_for_inactive_replica_timeout = 30, min_count_to_compile_expression = 0, compile_sort_description = false, group_by_two_level_threshold = 1000000, group_by_two_level_threshold_bytes = 1, distributed_aggregation_memory_efficient = false, allow_nonconst_timezone_arguments = true, max_parallel_replicas = 3, cluster_for_parallel_replicas = 'parallel_replicas', allow_experimental_parallel_reading_from_replicas = 1, parallel_replicas_for_non_replicated_merge_tree = true, input_format_parallel_parsing = false, min_chunk_bytes_for_parallel_parsing = 10162012, merge_tree_coarse_index_granularity = 24, min_bytes_to_use_direct_io = 10737418240, min_bytes_to_use_mmap_io = 10648023850, log_queries = true, insert_quorum_timeout = 60000, table_function_remote_max_addresses = 200, merge_tree_read_split_ranges_into_intersecting_and_non_intersecting_injection_probability = 0.4000000059604645, http_response_buffer_size = 4850398, fsync_metadata = false, http_send_timeout = 60., http_receive_timeout = 60., opentelemetry_start_trace_probability = 0.10000000149011612, prefer_external_sort_block_bytes = 0, max_bytes_before_external_sort = 10737418240, max_bytes_before_remerge_sort = 243713462, max_execution_time = 60., cross_join_min_rows_to_compress = 0, cross_join_min_bytes_to_compress = 1, max_memory_usage = 10000000000, max_memory_usage_for_user = 20000532480, max_untracked_memory = 1048576, memory_profiler_step = 1048576, log_comment = '03156_group_concat.sql', send_logs_level = 'fatal', prefer_localhost_replica = false, aggregation_in_order_max_block_bytes = 33373741, read_in_order_two_level_merge_threshold = 43, allow_introspection_functions = true, database_atomic_wait_for_drop_and_detach_synchronously = true, optimize_or_like_chain = true, optimize_if_transform_strings_to_enum = true, optimize_append_index = true, distributed_ddl_entry_format_version = 6, local_filesystem_read_method = 'mmap', local_filesystem_read_prefetch = true, merge_tree_compact_parts_min_granules_to_multibuffer_read = 119, async_insert_busy_timeout_max_ms = 5000, enable_filesystem_cache = true, enable_filesystem_cache_on_write_operations = true, throw_on_error_from_cache_on_write_operations = true, filesystem_cache_segments_batch_size = 10, load_marks_asynchronously = true, allow_prefetched_read_pool_for_remote_filesystem = false, allow_prefetched_read_pool_for_local_filesystem = false, filesystem_prefetch_step_bytes = 104857600, filesystem_prefetch_min_bytes_for_single_read_task = 8388608, filesystem_prefetch_max_memory_usage = 33554432, filesystem_prefetches_limit = 0, allow_deprecated_database_ordinary = true, optimize_sorting_by_input_stream_properties = false, insert_keeper_max_retries = 100, insert_keeper_retry_initial_backoff_ms = 1, insert_keeper_retry_max_backoff_ms = 1, insert_keeper_fault_injection_probability = 0.009999999776482582, ignore_drop_queries_probability = 0.5, optimize_distinct_in_order = false, session_timezone = 'Africa/Khartoum', prefer_warmed_unmerged_parts_seconds = 2
2024.06.18 21:16:13.351871 [ 3216 ] {} <Fatal> Application: Child process was terminated by signal 11.
```

### PR DESCRIPTION
**Pull request by yariks5s**
Closes #65426

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add aggregate function `groupConcat`. About the same as `arrayStringConcat( groupArray(column), ',')`
Can receive 2 parameters: a string delimiter and the number of elements to be processed

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 8d782ee09029c0f2240da7cdf2868818af7e3f11 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65451/8d782ee09029c0f2240da7cdf2868818af7e3f11/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/AggregateFunctions/AggregateFunctionGroupConcat.cpp [hunk 1]
```diff
@@ -178,10 +194,13 @@ class GroupConcatImpl final
     {
         auto & cur_data = this->data(place);
 
-        readVarUInt(cur_data.data_size, buf);
-        readVarUInt(cur_data.allocated_size, buf);
+        UInt64 temp_size;
```
Comment (CheSema) [inline src/AggregateFunctions/AggregateFunctionGroupConcat.cpp:197]:
```suggestion
        UInt64 temp_size = 0;
```

**Review by CheSema — APPROVED** [review]
As a fast fix it is just enough.
Lest add more tests and other improvements in different PR.

**Comment by Algunenano** [comment]
> As a fast fix it is just enough.
> Lest add more tests and other improvements in different PR.

Not enough. We should add a test so the fuzzer keeps iterating over the crash. Master should be stable and tested

**Comment by yariks5s** [comment]
@Algunenano repro of the fixed issue is created in `tests/queries/0_stateless/03195_group_concat_deserialization_fix.sql` as a separate file.

**Comment by Algunenano** [comment]
The fuzzer has found more problems, just because of the new tests. I'm going to revert the feature temporarily so we can branch 24.6, so please revert the revert here to continue

File: src/AggregateFunctions/AggregateFunctionGroupConcat.cpp [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    void deserialize(AggregateDataPtr __restrict place, ReadBuffer & buf, std::optional<size_t> /* version */, Arena * arena) const override
+    {
+        auto & cur_data = this->data(place);
+
+        UInt64 temp_size = 0;
+        readVarUInt(temp_size, buf);
+
+        cur_data.checkAndUpdateSize(temp_size, arena);
+
+        buf.readStrict(cur_data.data, temp_size);
```
Comment (CheSema) [inline src/AggregateFunctions/AggregateFunctionGroupConcat.cpp:202]:
```suggestion
        buf.readStrict(cur_data.data + cur_data.data_size, temp_size);
```

**Review by CheSema — APPROVED** [review]

File: src/AggregateFunctions/AggregateFunctionGroupConcat.cpp (hunk 2 above)
Comment (CheSema) [reply src/AggregateFunctions/AggregateFunctionGroupConcat.cpp:202]: Or we could add 
```
chassert(cur_data.data_size == 0);
```
instead.

Comment (CheSema) [reply src/AggregateFunctions/AggregateFunctionGroupConcat.cpp]: I think there is no problem now, but this is to make it for sure.

**Comment by yariks5s** [comment]
- Stateless Tests (tsan, s3 Storage) [2/5]: 01055_window_view_proc_hop_to, 00763_lock_buffer_long, 01171_mv_select_insert_isolation_long - not related, 2 failed because of timeouts
- Stateless tests (tsan, s3 storage) [1/5]: 00988_parallel_parts_removal - broken test
- Stateless Tests (msan) [6/6]: 02867_nullable_primary_key_final, 00871_t64_codec_signed - not related, timeout
- Stateless Tests (debug) [4/5]: 00626_replace_partition_from_table_zookeeper - not related, timeout
- Integration Tests (tsan) [4/6]: test_materialized_database_mysql_drop_ddl - not related

**Comment by Algunenano** [comment]
Please remember to update the changelog entry to a new feature

**Pull request merged**